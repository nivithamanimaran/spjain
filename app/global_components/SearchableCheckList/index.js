import { Grid, InputBase, makeStyles, useTheme } from '@material-ui/core';
import React, { useEffect, useState } from 'react';
import SearchIcon from '@material-ui/icons/Search';
import KenRadioGroup from '../KenRadioGroup';
import KenCheckboxGroup from '../KenCheckboxGroup';

const useStyles = makeStyles(theme => ({
  formGroup: {
    flexDirection: 'column',
  },
  formControlLabelRoot: {
    marginLeft: '-8px',
  },
  searchBarContainer: {
    display: 'flex',
    alignItems: 'center',
    padding: '5px',
    border: `1px solid ${theme.palette.KenColors.assessmentBorder}`,
    borderRadius: '3px',
    marginBottom: '5px',
  },
  checkListSection: {
    '&::-webkit-scrollbar': {
      width: '4px',
    },
    '&::-webkit-scrollbar-track': {
      borderRadius: '10px',
      background: theme.palette.KenColors.scrollbarColor,
    },
    '&::-webkit-scrollbar-thumb': {
      borderRadius: '10px',
      backgroundColor: `${theme.palette.KenColors.neutral103}`,
    },
  },
}));

const SearchableCheckList = props => {
  const {
    isSearchEnabled = false,
    searchTextValue = '',
    type,
    name,
    value, // for checkbox an array of values to be checked, for radio buttons a single string value to be selected
    options, // an array of Objects({label: "", value: ""})
    handleChange,
    isSearchCaseSensitive = false,
    checkSectionStyleClass,
  } = props;

  const classes = useStyles();

  const [checkListItems, setCheckListItems] = useState([]);
  const [searchText, setSearchText] = useState(searchTextValue);

  useEffect(() => {
    let tempList = [...options];
    if (tempList.length) {
      if (searchText.length && isSearchEnabled) {
        tempList = tempList.filter(check => {
          return isSearchCaseSensitive
            ? check.label.includes(searchText)
            : check.label.toLowerCase().includes(searchText.toLowerCase());
        });

        tempList = tempList.map(option => ({
          ...option,
          label: getHighlightedText(
            option.label,
            searchText,
            isSearchCaseSensitive
          ),
        }));
      }
    }

    setCheckListItems(tempList);
  }, [options, searchText]);

  const theme = useTheme();
  return (
    <>
      <Grid container data-testid="checklist-container">
        {isSearchEnabled && (
          <Grid item xs={12} data-testid="search-bar">
            <div className={classes.searchBarContainer}>
              <SearchIcon
                style={{ color: theme.palette.KenColors.neutral100 }}
              />
              <InputBase
                size="small"
                placeholder="Search"
                inputProps={{ 'data-testid': 'search-input' }}
                value={searchText}
                onChange={e => setSearchText(e.target.value)}
              />
            </div>
          </Grid>
        )}

        <Grid
          item
          xs={12}
          className={`${classes.checkListSection} ${checkSectionStyleClass}`}
        >
          {type === 'checkbox' ? (
            <span data-testid="checkbox-group">
              <KenCheckboxGroup
                value={value}
                options={checkListItems}
                onChange={handleChange}
              />
            </span>
          ) : null}
          {type === 'radio' ? (
            <span data-testid="radio-group">
              <KenRadioGroup
                name={name}
                label={''}
                value={value}
                options={checkListItems}
                onChange={handleChange}
                row={false}
                classOverrides={{
                  formControlLabelRoot: classes.formControlLabelRoot,
                }}
              />
            </span>
          ) : null}
        </Grid>
      </Grid>
    </>
  );
};

const getHighlightedText = (text, highlight, isSearchCaseSensitive) => {
  if (highlight && highlight.length) {
    let strArr = [];
    let tempText;
    if (isSearchCaseSensitive) {
      tempText = text;
    } else {
      tempText = text.toLowerCase();
      highlight = highlight.toLowerCase();
    }

    let index = 0;
    let startIndex = 0;
    while (index !== -1) {
      index = tempText.indexOf(highlight, startIndex);
      if (index !== -1) {
        strArr.push(text.substring(startIndex, index));
        strArr.push(text.substring(index, index + highlight.length));
      } else {
        strArr.push(text.substring(startIndex, text.length));
      }
      startIndex = index + highlight.length;
    }
    return (
      <>
        {strArr.map(item => {
          if (
            (isSearchCaseSensitive === false &&
              item.toLowerCase() === highlight.toLowerCase()) ||
            (isSearchCaseSensitive === true && item === highlight)
          ) {
            return <b>{item}</b>;
          } else {
            return <>{item}</>;
          }
        })}
      </>
    );
  } else {
    return text;
  }
};

export default SearchableCheckList;
