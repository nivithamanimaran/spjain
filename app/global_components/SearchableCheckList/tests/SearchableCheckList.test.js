import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import '@testing-library/jest-dom/extend-expect';
import ReactDOM from 'react-dom';
import ThemeProvider from '../../../utils/themeProvider';
import { BrowserRouter as Router } from 'react-router-dom';
import SearchableCheckList from '../index';

afterEach(() => {
  cleanup();
  jest.clearAllMocks();
});
const options = [
  { label: 'Item 1', value: 'item1' },
  { label: 'Item 2', value: 'item2' },
  { label: 'Item 3', value: 'item3' },
];
const handleChange = jest.fn();
describe('Searchable check list', () => {
  it('should render and match the snapshot for checkbox', () => {
    const { getByTestId, queryByTestId } = render(
      <ThemeProvider>
        <Router>
          <SearchableCheckList
            isSearchEnabled={true}
            type={'checkbox'}
            name={''}
            value={[]}
            options={options}
            handleChange={handleChange}
            isSearchCaseSensitive={false}
          />
        </Router>
      </ThemeProvider>
    );

    expect(getByTestId('checklist-container')).toMatchSnapshot();
    expect(queryByTestId('checkbox-group')).toBeTruthy();
    expect(queryByTestId('radio-group')).toBeFalsy();
  });
  it('should render and match the snapshot for radio button', () => {
    const { getByTestId, queryByTestId } = render(
      <ThemeProvider>
        <Router>
          <SearchableCheckList
            isSearchEnabled={true}
            type={'radio'}
            name={''}
            value={[]}
            options={options}
            handleChange={handleChange}
            isSearchCaseSensitive={false}
          />
        </Router>
      </ThemeProvider>
    );

    expect(getByTestId('checklist-container')).toMatchSnapshot();
    expect(queryByTestId('checkbox-group')).toBeFalsy();
    expect(queryByTestId('radio-group')).toBeTruthy();
  });

  it('Search functionality should work', () => {
    const { container, queryByTestId } = render(
      <ThemeProvider>
        <Router>
          <SearchableCheckList
            isSearchEnabled={true}
            type={'checkbox'}
            name={''}
            value={[]}
            options={options}
            handleChange={handleChange}
            isSearchCaseSensitive={false}
          />
        </Router>
      </ThemeProvider>
    );

    let checks = container.querySelectorAll("input[type='checkbox']");
    expect(checks).toHaveLength(3);
    const input = queryByTestId('search-input');
    fireEvent.change(input, { target: { value: 'item 1' } });
    expect(input).toHaveValue('item 1');
    checks = container.querySelectorAll("input[type='checkbox']");
    expect(checks).toHaveLength(1);
  });

  it('default Search text value functionality should work', () => {
    const { container } = render(
      <ThemeProvider>
        <Router>
          <SearchableCheckList
            isSearchEnabled={true}
            searchTextValue="item 1"
            type={'checkbox'}
            name={''}
            value={[]}
            options={options}
            handleChange={handleChange}
            isSearchCaseSensitive={false}
          />
        </Router>
      </ThemeProvider>
    );

    let checks = container.querySelectorAll("input[type='checkbox']");
    expect(checks).toHaveLength(1);
    expect(container.querySelector("input[name='item1']")).toBeTruthy();
    expect(container.querySelector("input[name='item2']")).toBeFalsy();
    expect(container.querySelector("input[name='item3']")).toBeFalsy();
  });

  it('Verify handleChange is called', () => {
    const { container } = render(
      <ThemeProvider>
        <Router>
          <SearchableCheckList
            isSearchEnabled={true}
            searchTextValue=""
            type={'checkbox'}
            name={''}
            value={[]}
            options={options}
            handleChange={handleChange}
            isSearchCaseSensitive={false}
          />
        </Router>
      </ThemeProvider>
    );

    fireEvent.click(container.querySelector('input[name="item1"]'));
    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toHaveBeenCalledWith(['item1']);
    cleanup();
    const { container: container1 } = render(
      <ThemeProvider>
        <Router>
          <SearchableCheckList
            isSearchEnabled={true}
            searchTextValue=""
            type={'checkbox'}
            name={''}
            value={['item1']}
            options={options}
            handleChange={handleChange}
            isSearchCaseSensitive={false}
          />
        </Router>
      </ThemeProvider>
    );
    fireEvent.click(container1.querySelector('input[name="item2"]'));
    expect(handleChange).toHaveBeenCalledTimes(2);
    expect(handleChange.mock.calls[1][0]).toEqual(['item1', 'item2']);
  });

  it('Verify handleChange is called after search', () => {
    const { container } = render(
      <ThemeProvider>
        <Router>
          <SearchableCheckList
            isSearchEnabled={true}
            searchTextValue=""
            type={'checkbox'}
            name={''}
            value={[]}
            options={options}
            handleChange={handleChange}
            isSearchCaseSensitive={false}
          />
        </Router>
      </ThemeProvider>
    );

    fireEvent.click(container.querySelector('input[name="item1"]'));
    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toHaveBeenCalledWith(['item1']);

    const { container: container1, queryByTestId } = render(
      <ThemeProvider>
        <Router>
          <SearchableCheckList
            isSearchEnabled={true}
            searchTextValue=""
            type={'checkbox'}
            name={''}
            value={['item1']}
            options={options}
            handleChange={handleChange}
            isSearchCaseSensitive={false}
          />
        </Router>
      </ThemeProvider>
    );

    const input = queryByTestId('search-input');
    fireEvent.change(input, { target: { value: 'item 2' } });
    fireEvent.click(container1.querySelector('input[name="item2"]'));
    expect(handleChange).toHaveBeenCalledTimes(2);
    expect(handleChange.mock.calls[1][0]).toEqual(['item1', 'item2']);
  });
  it('Search functionality should work for partial match', () => {
    const { container, queryByTestId } = render(
      <ThemeProvider>
        <Router>
          <SearchableCheckList
            isSearchEnabled={true}
            type={'checkbox'}
            name={''}
            value={[]}
            options={[...options, { label: 'Test data', value: 'test' }]}
            handleChange={handleChange}
            isSearchCaseSensitive={false}
          />
        </Router>
      </ThemeProvider>
    );

    let checks = container.querySelectorAll("input[type='checkbox']");
    expect(checks).toHaveLength(4);
    const input = queryByTestId('search-input');
    fireEvent.change(input, { target: { value: 'item' } });
    expect(input).toHaveValue('item');
    checks = container.querySelectorAll("input[type='checkbox']");
    expect(checks).toHaveLength(3);
    checks.forEach((check, index) => {
      expect(check.getAttribute('name')).toBe(options[index].value);
    });
  });
});
