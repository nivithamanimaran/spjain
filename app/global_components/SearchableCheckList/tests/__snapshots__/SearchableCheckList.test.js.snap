// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchable check list should render and match the snapshot for checkbox 1`] = `
<div
  class="MuiGrid-root MuiGrid-container"
  data-testid="checklist-container"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    data-testid="search-bar"
  >
    <div
      class="makeStyles-searchBarContainer-3"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        style="color: rgb(122, 134, 154);"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <div
        class="MuiInputBase-root"
      >
        <input
          class="MuiInputBase-input"
          data-testid="search-input"
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root makeStyles-checkListSection-4 undefined MuiGrid-item MuiGrid-grid-xs-12"
  >
    <span
      data-testid="checkbox-group"
    >
      <div
        data-testid="checkbox-group"
      >
        <div
          class="MuiFormGroup-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
            style="flex-wrap: nowrap;"
          >
            <label
              class="MuiFormControlLabel-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiCheckbox-root Component-root-7 makeStyles-transparentBg-6"
                component-root="[object Object]"
                style="margin-right: -20px;"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    class="PrivateSwitchBase-input-12"
                    data-indeterminate="false"
                    name="item1"
                    type="checkbox"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              />
            </label>
            <p
              class="MuiTypography-root MuiTypography-body1"
              style="font-size: 12px; color: rgb(80, 95, 121);"
            >
               
              Item 1
               
            </p>
          </div>
        </div>
        <div
          class="MuiFormGroup-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
            style="flex-wrap: nowrap;"
          >
            <label
              class="MuiFormControlLabel-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiCheckbox-root Component-root-7 makeStyles-transparentBg-6"
                component-root="[object Object]"
                style="margin-right: -20px;"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    class="PrivateSwitchBase-input-12"
                    data-indeterminate="false"
                    name="item2"
                    type="checkbox"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              />
            </label>
            <p
              class="MuiTypography-root MuiTypography-body1"
              style="font-size: 12px; color: rgb(80, 95, 121);"
            >
               
              Item 2
               
            </p>
          </div>
        </div>
        <div
          class="MuiFormGroup-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
            style="flex-wrap: nowrap;"
          >
            <label
              class="MuiFormControlLabel-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiCheckbox-root Component-root-7 makeStyles-transparentBg-6"
                component-root="[object Object]"
                style="margin-right: -20px;"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    class="PrivateSwitchBase-input-12"
                    data-indeterminate="false"
                    name="item3"
                    type="checkbox"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              />
            </label>
            <p
              class="MuiTypography-root MuiTypography-body1"
              style="font-size: 12px; color: rgb(80, 95, 121);"
            >
               
              Item 3
               
            </p>
          </div>
        </div>
      </div>
    </span>
  </div>
</div>
`;

exports[`Searchable check list should render and match the snapshot for radio button 1`] = `
<div
  class="MuiGrid-root MuiGrid-container"
  data-testid="checklist-container"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    data-testid="search-bar"
  >
    <div
      class="makeStyles-searchBarContainer-15"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        style="color: rgb(122, 134, 154);"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <div
        class="MuiInputBase-root"
      >
        <input
          class="MuiInputBase-input"
          data-testid="search-input"
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root makeStyles-checkListSection-16 undefined MuiGrid-item MuiGrid-grid-xs-12"
  >
    <span
      data-testid="radio-group"
    >
      <div
        class="MuiFormGroup-root makeStyles-formGroup-21 makeStyles-formGroup-21"
      >
        <fieldset
          class="MuiFormControl-root"
        >
          
          <div
            aria-label=""
            class="MuiFormGroup-root"
            role="radiogroup"
          >
            <div>
              <label
                class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-14"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-22 MuiRadio-root makeStyles-root-17"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-25"
                      name=""
                      type="radio"
                      value="item1"
                    />
                    <span
                      class="makeStyles-icon-19"
                    />
                  </span>
                </span>
                <span
                  class="MuiTypography-root MuiFormControlLabel-label makeStyles-labelItem-18 MuiTypography-body1"
                >
                  Item 1
                </span>
              </label>
            </div>
            <div>
              <label
                class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-14"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-22 MuiRadio-root makeStyles-root-17"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-25"
                      name=""
                      type="radio"
                      value="item2"
                    />
                    <span
                      class="makeStyles-icon-19"
                    />
                  </span>
                </span>
                <span
                  class="MuiTypography-root MuiFormControlLabel-label makeStyles-labelItem-18 MuiTypography-body1"
                >
                  Item 2
                </span>
              </label>
            </div>
            <div>
              <label
                class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-14"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-22 MuiRadio-root makeStyles-root-17"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-25"
                      name=""
                      type="radio"
                      value="item3"
                    />
                    <span
                      class="makeStyles-icon-19"
                    />
                  </span>
                </span>
                <span
                  class="MuiTypography-root MuiFormControlLabel-label makeStyles-labelItem-18 MuiTypography-body1"
                >
                  Item 3
                </span>
              </label>
            </div>
          </div>
        </fieldset>
      </div>
    </span>
  </div>
</div>
`;
